tests/test_dataset.py

```diff
@@ -0,0 +1,520 @@
+    ds.add_layer("color", "color")
+
+    ds.get_layer("color").add_mag("1")
+    ds.get_layer("color").add_mag("2-2-1")
+
+    assert path.exists("../testoutput/TiffDataset/color/1")
+    assert path.exists("../testoutput/TiffDataset/color/2-2-1")
+
+    assert len(ds.properties.data_layers) == 1
+    assert len(ds.properties.data_layers["color"].wkw_resolutions) == 2
+
+
+def test_open_wk_dataset():
+    ds = WKDataset.open("../testdata/simple_wk_dataset")
+
+    assert len(ds.properties.data_layers) == 1
+    assert len(ds.properties.data_layers["color"].wkw_resolutions) == 1
+
+
+def test_open_tiff_dataset():
+    ds = TiffDataset.open("../testdata/simple_tiff_dataset")
+
+    assert len(ds.properties.data_layers) == 1
+    assert len(ds.properties.data_layers["color"].wkw_resolutions) == 1
+
+
+def test_slice_read_with_open():
+    # This test would be the same for TiffDataset
+
+    wk_slice = WKDataset.open("../testdata/simple_wk_dataset/").get_slice("color", "1")
+
+    assert not wk_slice._is_opened
+
+    with wk_slice.open():
+        assert wk_slice._is_opened
+
+        data = wk_slice.read((10, 10, 10))
+        assert data.shape == (3, 10, 10, 10)  # three channel
+
+    assert not wk_slice._is_opened
+
+
+def test_slice_read_without_open():
+    # This test would be the same for TiffDataset
+
+    wk_slice = WKDataset.open("../testdata/simple_wk_dataset/").get_slice("color", "1")
+
+    assert not wk_slice._is_opened
+
+    # 'read()' checks if it was already opened. If not, it opens and closes automatically
+    data = wk_slice.read((10, 10, 10))
+    assert data.shape == (3, 10, 10, 10)  # three channel
+
+    assert not wk_slice._is_opened
+
+
+def test_slice_wk_write():
+    delete_dir("../testoutput/simple_wk_dataset/")
+    copytree("../testdata/simple_wk_dataset/", "../testoutput/simple_wk_dataset/")
+
+    wk_slice = WKDataset.open("../testoutput/simple_wk_dataset/").get_slice(
+        "color", "1", size=(100, 100, 100)
+    )
+
+    with wk_slice.open():
+        np.random.seed(1234)
+        write_data = np.random.rand(3, 10, 10, 10).astype(np.uint8)
+
+        wk_slice.write(write_data)
+
+        data = wk_slice.read((10, 10, 10))
+        assert np.array_equal(data, write_data)
+
+
+def test_slice_tiff_write():
+    delete_dir("../testoutput/simple_tiff_dataset/")
+    copytree("../testdata/simple_tiff_dataset/", "../testoutput/simple_tiff_dataset/")
+
+    tiff_slice = TiffDataset.open("../testoutput/simple_tiff_dataset/").get_slice(
+        "color", "1", size=(100, 100, 100)
+    )
+
+    with tiff_slice.open():
+        np.random.seed(1234)
+        write_data = np.random.rand(5, 5, 5).astype(np.uint8)
+
+        tiff_slice.write(np.zeros((5, 5, 5), dtype=np.uint8))
+
+        data = tiff_slice.read((5, 5, 5))
+        assert data.shape == (1, 5, 5, 5)  # this dataset has only one channel
+        assert np.array_equal(data, np.expand_dims(write_data, 0))
```

Comment:
How can `write_data be equal to `data` when `write_data` wasn't written to the dataset in this test? Does it rely on a previous test?

Reply:
Good catch. The random values of `np.random.rand` are between 0 and 1. Converting them to uint8 results in zeros. This was also the case in other tests.